### DISCUSSION TITLE
Can't get a secondary handle to a TcpStream

### DISCUSSION THREAD

**Discussion by CenTdemeern1** [rust3ds/ctru-rs#202] [Q&A]
Using `TcpStream::try_clone()` always nets `Err(Os { code: 22, kind: InvalidInput, message: "Invalid argument" })`. Am I doing something wrong, or is this a bug?

**Comment by ian-h-chamberlain** [comment]
Hmm, this might be a limitation of `libctru`... From what I can gather, `try_clone()` ultimately calls into this code to clone the file descriptor of the underlying socket: https://github.com/rust-lang/rust/blob/f33a8c6426074b7ce8d08740e9805fdca96ee150/library/std/src/os/fd/owned.rs#L101-L118

However, the `fcntl` implementation doesn't seem to support `F_DUPFD` (or similar) as a command:
https://github.com/devkitPro/libctru/blob/30dc65121dc37ad5faf33ca842fd8d63f425f04f/libctru/source/services/soc/soc_fcntl.c#L48-L51

Without upstream support, we'd probably be better off changing Rust `std` to return `UNSUPPORTED_PLATFORM` here, or possibly changing the implementation to use one of the `dup` functions, which I believe _is_ supported by `libctru`? Either that, or upstreaming proper `F_DUPFD` support to libctru.

In the meantime, as a workaround, you can try using the `dup` calls yourself, I'm guessing it'd look something like this:
```rs
let fd = stream.as_raw_fd();
let new_stream = unsafe {
	let new_fd = libc::dup(fd);
	if new_fd < 0 {
		// handle error
	}
	TcpStream::from_raw_fd(new_fd)
};
```
